### PR TITLE
[message-tags] Drop 3.3 and merge in 3.2

### PR DESCRIPTION
**Pull request by jwheare**
[Rendered view](https://github.com/jwheare/ircv3-specifications/blob/message-tags-merged/extensions/message-tags.md)

* [x] Drop 3.3 and move to extensions
* [x] Merge 3.2 in to make a single document
* [x] Make sure old URLs redirect

## Known implementations

Unchecked means incomplete/unmerged or an intent to implement has been expressed. Any others?

### Server (2/2)

* [x] IRCCloud teams (@jwheare)
* [x] oragono (@DanielOaks) https://oragono.io/specs.html
* [x] Inspircd (@SaberUK) https://github.com/SaberUK/inspircd/compare/f5de558...master%2Bircv3-draft

### Client (2/2)

* [x] IRCCloud (@jwheare)
* [x] kiwiirc (@prawnsalad)

### Bots

* [x] bitbot (@jesopo)

### Libraries

* [x] irc-framework (@prawnsalad)
* [x] girc (@lrstanley)

### CODE REVIEW

**Comment by DanielOaks** [comment]
Noice, looks good and excited for the merging. As a reminder, when pushing stuff like this through make sure to not update the web side (esp as we have some stuff pending that'll require additional work to get merged in properly on that side).

**Comment by SaberUK** [comment]
This should probably include a redirect for the old URL?

**Comment by jwheare** [comment]
Added redirects and merged the two documents. This also drops the draft/wip status and sets the final cap name to `message-tags`.

We can either ratify the new message-tags cap first before merging, or use this PR to do it.

File: extensions/message-tags.md [hunk 1]
```diff
@@ -1,97 +1,143 @@
...
+| all others      | the character itself          |
+
+This escape format is more space efficient than URL-escaping, and ensures that message parts can easily by split on spaces and semi-colons before further parsing.
+
+If a lone `\` exists at the end of an escaped value (with no escape character following it), then there
+SHOULD be no output character. For example, the escaped value `test\` should unescape to `test`.
+
+## Size limit
+
+The size limit for message tags is 4607 bytes, including the leading `'@'` (0x40) and trailing space `' '` (0x20) characters. The size limit for the rest of the message remains unchanged at 512 bytes.
```
[the remark this answers is not in the document]
Comment (DanielOaks) [reply extensions/message-tags.md]: If this is supposed to act as it does right now, i.e. if clients enable the `message-tags` cap then the limit gets boosted to `4607`, otherwise it stays at the standard `512`, then it should be explicitly mentioned that's how it works in the spec.

File: extensions/message-tags.md [hunk 2]
```diff
@@ -101,37 +147,44 @@ A new message command `TAGMSG` is defined for sending messages with tags but no
...
+e.g. `xn--e1afmkfd.org/foo`.
 
-Servers MUST reply with the `ERR_INPUTTOOLONG` (`417`) error numeric if a client sends a message with more tag data than the allowed limit. Servers MUST NOT truncate tags and MUST always reject lines that exceed this limit.
+Vendor-Specific tags should be submitted to the IRCv3 working group for consideration.
 
-    417    ERR_INPUTTOOLONG
-          ":Input line was too long"
+### Standardized
+
+Names for which a corresponding document sits in the IRCv3 Extension Registry.
```
Comment (DanielOaks) [inline extensions/message-tags.md]: this should link to the `[IRCv3 Tag Registry](https://ircv3.net/registry.html#tags)`

File: extensions/message-tags.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (DanielOaks) [reply extensions/message-tags.md]: and/or if we are changing behaviour explicitly, this should be listed in the errata of the spec

File: extensions/message-tags.md [hunk 3]
```diff
@@ -1,97 +1,143 @@
...
+## Introduction
 
-This is a work-in-progress specification.
+Message tags are a mechanism for adding additional metadata on a per-message basis. This is achieved via an extension to the protocol message format, enabled via capability negotiation.
 
-Software implementing this work-in-progress specification MUST NOT use the
-unprefixed `message-tags` capability name. Instead, implementations SHOULD use
-the `draft/message-tags-0.2` capability name to be interoperable with other software
-implementing a compatible work-in-progress version.
+Tagged messages can be initiated by both servers and clients. The usage of individual tags are specified in their own documents.
```
[the remark this answers is not in the document]
Comment (jwheare) [reply extensions/message-tags.md]: Well a client initiated tag can still be sent by a server, the phrase is used elsewhere to clarify client only tags. But yeah maybe worth the eli5 treatment.

File: extensions/message-tags.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (jwheare) [reply extensions/message-tags.md]: For tags that aren't using the additional CAP, you'll never reach the 512 limit, so it doesn't make much practical difference. You won't ever be sent more than that if you don't negotiate it.

[the remark this answers is not in the document]
Comment (jwheare) [reply extensions/message-tags.md]: The *only* issue I can see here is current implementations maybe worrying they might not be “compliant” if they do any sort of limit checking as a hard coded 512. But practically it won’t ever be an issue for them.

Also, the new wording covers them fine: server tags are still limited to 512, and that’s all they’ll ever see.

[the remark this answers is not in the document]
Comment (DanielOaks) [reply extensions/message-tags.md]: Hmm, yeah we are gonna have those grandfathered impls that don't comply with the spec exactly as written (_"Clients and servers that negotiate a capability that uses tags MUST support the full tag specification"_ they don't support this size limit part as written). Not gonna be an issue for them, but still worth keeping in mind as we push this forwards. I don't have a big drama with the line above or requiring it for future tag capabilities though, feels like it should be fine, would appreciate input from other server implementers on the change.

[the remark this answers is not in the document]
Comment (jwheare) [reply extensions/message-tags.md]: Again, it’s not a change :)

[the remark this answers is not in the document]
Comment (DanielOaks) [reply extensions/message-tags.md]: It is a change from the [stock IRCv3.2 message tags spec](https://ircv3.net/specs/core/message-tags-3.2.html) used by account-tag and server-time. It's not an impactful change or one that will break running servers but it is a change. With just that spec, tag space by the spec is `512`. With this new spec, tag space by the spec is `4094+512`. It doesn't actively impact anything since the servers implementing just those tags aren't gonna hit the limit, but they will not comply 100% with the spec (and it's important to keep that in mind so we don't run into something like the Metadata spec and get any backlash around shifting compatibility/requirements).

[the remark this answers is not in the document]
Comment (jwheare) [reply extensions/message-tags.md]: It’s a zen change. If a tree falls in the wood with no one around to hear it type thing. Not worth worrying about or worrying about other people worrying about :)

[the remark this answers is not in the document]
Comment (DanielOaks) [reply extensions/message-tags.md]: Sure, just keep it in mind

File: extensions/message-tags.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (jwheare) [reply extensions/message-tags.md]: Dropped the use of initiated everywhere.

File: extensions/message-tags.md [hunk 4]
```diff
@@ -1,97 +1,143 @@
...
+    <tag>           ::= <key> ['=' <escaped value>]
+    <key>           ::= [ '+' ] [ <vendor> '/' ] <sequence of letters, digits, hyphens ('-')>
+    <escaped value> ::= <sequence of any characters except NUL, CR, LF, semicolon (`;`) and SPACE>
+    <vendor>        ::= <host>
 
-* Adds a new capability to indicate base support for message tags
-* Defines a prefix for expressing client-only tags
-* Defines a new command for tag-only messages
-* Increases the size limit for tags
+Not all tag keys have a value.
```
[the remark this answers is not in the document]
Comment (jwheare) [reply extensions/message-tags.md]: Added an explicit explanation of key values.

File: extensions/message-tags.md [hunk 5]
```diff
@@ -1,97 +1,143 @@
...
+| Character       | Sequence in `<escaped value>` |
+|-----------------|-------------------------------|
+| `;` (semicolon) | `\:` (backslash and colon)    |
+| `SPACE`         | `\s`                          |
+| `\`             | `\\`                          |
+| `CR`            | `\r`                          |
+| `LF`            | `\n`                          |
+| all others      | the character itself          |
+
+This escape format is more space efficient than URL-escaping, and ensures that message parts can easily by split on spaces and semi-colons before further parsing.
```
[the remark this answers is not in the document]
Comment (jwheare) [reply extensions/message-tags.md]: Dropped.

File: extensions/message-tags.md (hunk 2 above)
Comment (jwheare) [reply extensions/message-tags.md]: Done

**Comment by jwheare** [comment]
That should cover it, thanks for the comments.

**Comment by DanielOaks** [comment]
Perfect, great changes. +1 from me

**Comment by jwheare** [comment]
I still would like to do a final set of tests with several clients and servers. Can do that when I’m back from holiday next week. Is an insp testnet running with this implementation?

**Comment by DanielOaks** [comment]
Yeah, let's get this merged in.

**Comment by SaberUK** [comment]
We are ready to ship it as soon as the spec is merged. 👍

**Comment by jwheare** [comment]
Does anyone have thoughts on increasing the server side limit for tag data. Reminder:

```
<server_max>    (512)  :: '@' <tag_data  510> ' '
<client_max>   (4096)  :: '@' <tag_data 4094> ' '
<combined_max> (4607)  :: '@' <tag_data  510> ';' <tag_data 4094> ' '
```

We could bump server_max up to 4096. The initial reason to restrict it was mostly a lack of imagination, IIRC. Client tags were seen as the place where things like long JSON blobs of attachment data would be sent. But I could see a case where some servers (especially gateways to other services) could emit such client tags in a server-validated form. That would mean that anything sent as a client tag might well be forwarded in the server tag section.

Any reason we might not want to do this? A larger limit just feels a bit more future proof to me.

**Comment by jwheare** [comment]
(btw, tested the InspIRCd tags implementation, seems to all be working well now, so I think we're now good for implementations)

**Comment by SaberUK** [comment]
:shipit:

**Comment by jwheare** [comment]
@DanielOaks @prawnsalad please can you chime in any thoughts as implementers on extending the server_max length?

**Comment by DanielOaks** [comment]
@jwheare first thought that pops into my head is to maybe only let servers send more than the `512` to clients that've negotiated `message-tags`, specifically because some older s/w might crash if it gets that much data. buuuuut. Clients that fall into that case won't have a chance of coming to that 512 limit because they won't have enough caps with tags negotiated with the server in the first place.

Up the server_max limit too, makes it consistent across the board. Clients will have a chance to extend their message length parsing as they implement more caps and the `message-tags` cap in particular, no special stuff needed.

**Comment by jwheare** [comment]
OK, increased the server limit. This is the last call for comments.

File: extensions/message-tags.md [hunk 6]
```diff
@@ -1,96 +1,123 @@
...
-work-in-progress: true
-updates:
-  - message-tags-3.2
+redirect_from:
+  - /specs/core/message-tags-3.3.html
+  - /specs/core/message-tags-3.2.html
 copyrights:
+  -
+    name: "Alexey Sokolov"
+    period: 2012-2014
```
Comment (Zarthus) [inline extensions/message-tags.md]: missing quotes

**Comment by jwheare** [comment]
No more comments, so this will merge tomorrow.

**Comment by jwheare** [comment]
This was published to the website in https://github.com/ircv3/ircv3.github.io/commit/0d5ed78b5b73eca4d00938cca6155d1570f63255
